### ISSUE TITLE
Rethink dependency auto-install during Prisma Client's generation

### ISSUE THREAD

**Issue by madebysid** [prisma/specs#451]
I'm specifically talking about this part of the Prisma Client spec: https://github.com/prisma/specs/tree/master/cli#install-prismaclient

I'm told that the reasoning behind introducing it was to smoothen the first user experience, but I urge you to rethink our solution.

Here are the problems I see currently:

1. This smells like an anti-pattern. `prisma2 generate` is touching and modifying files outside its own "scope". This feels like an abuse of the lack of a permissions system in Node. It is not `prisma2`'s responsibility to manage dependencies. In less technical terms, it is too "magical".

2. We're making assumptions about the end user's setup. They might be working with yarn/npm workspaces, might have other symlinks in their `node_modules` that will be disturbed by a `yarn install` command, or they might not even be using `npm` or `yarn` to manage dependencies at all (think `pnpm`, `lerna` or `yarn2`) The point is, we don't have enough information to figure all this out, and we shouldn't need to. As an example: In Studio's tests, I use the SDK to generate the Client (in other words, the programmatic API of Prisma Client).

3. This is not CI / test friendly. If I run `prisma2 generate` during a test, it will have a side-effect of modifying my `package.json` after it has run. This might not be as significant, but it highlights the fact that our current solution does not take a lot of use cases into account.

Here are some solutions I propose. Please correct me if I'm missing some context:

1. When I run `prisma2 generate` in a project where we're able to find a `schema.prisma` file but the generator cannot be resolved, error out. If the error clearly tells me that I need to install a separate package to use the `generator` I have defined in my `prisma.schema`, I do not see it as a friction point in our "Getting Started" experience.

2. Instead of going back to the old way of running `prisma2 generate` in `postinstall` hooks, we could show a warning in `development` mode about Prisma Client being "out of sync" when it detects a `schema.prisma` change. This could happen during the Client's first `connect()` call, which happens automatically.
(AFAICT, a nasty side-effect of a failed `postinstall` script in a package is that the user cannot even `npm i` without getting rid of the faulty package, which is why I'm against a `postinstall` script in the first place.)
We might also be able to look at tools like Prettier and run `prisma2 generate` whenever schema.prisma changes as part of our VS Code plugin. This obviously won't work for people on Sublime etc., though.

**Comment by janpio** [comment]
> 1. I do not see it as a friction point in our "Getting Started" experience.

Wouldn't that turn 1 command into 3 commands?

**Comment by madebysid** [comment]
Yes indeed, but my line of thought is that those would be "guided" steps. It'd make things explicit and teach the user exactly what's happening.

**Comment by schickling** [comment]
I agree with most points you've presented @madebysid. We'll look into this once Prisma 2.0 is out of the door.
